SmartDeviceLink/SDLMenuManager.m

```diff
@@ -0,0 +1,400 @@
+        _waitingOnHMILevelUpdate = YES;
+        _menuCells = menuCells;
+        return;
+    }
+
+    // Check for duplicate titles
+    NSMutableSet *titleCheckSet = [NSMutableSet set];
+    for (SDLMenuCell *cell in menuCells) {
+        [titleCheckSet addObject:cell.title];
+    }
+    if (titleCheckSet.count != menuCells.count) {
+        SDLLogE(@"Not all cell titles are unique. The menu will not be set.");
+        return;
+    }
+
+    // Set the ids
+    self.lastMenuId = 0;
+    [self sdl_updateIdsOnMenuCells:menuCells parentId:ParentIdNotFound];
+
+    _oldMenuCells = _menuCells;
+    _menuCells = menuCells;
+
+    // Upload the artworks
+    NSArray<SDLArtwork *> *artworksToBeUploaded = [self sdl_findAllArtworksToBeUploadedFromCells:self.menuCells];
+    if (artworksToBeUploaded.count > 0) {
+        [self.fileManager uploadArtworks:artworksToBeUploaded completionHandler:^(NSArray<NSString *> * _Nonnull artworkNames, NSError * _Nullable error) {
+            if (error != nil) {
+                SDLLogE(@"Error uploading menu artworks: %@", error);
+            }
+
+            SDLLogD(@"Menu artworks uploaded");
+            [self sdl_updateWithCompletionHandler:nil];
+        }];
+    }
+
+    [self sdl_updateWithCompletionHandler:nil];
+}
+
+#pragma mark - Helpers
+
+#pragma mark Artworks
+
+- (NSArray<SDLArtwork *> *)sdl_findAllArtworksToBeUploadedFromCells:(NSArray<SDLMenuCell *> *)cells {
+    if (![self.displayCapabilities hasImageFieldOfName:SDLImageFieldNameCommandIcon]) {
+        return @[];
+    }
+
+    NSMutableArray<SDLArtwork *> *mutableArtworks = [NSMutableArray array];
+    for (SDLMenuCell *cell in cells) {
+        if (cell.icon != nil && ![self.fileManager hasUploadedFile:cell.icon]) {
+            [mutableArtworks addObject:cell.icon];
+        }
+
+        if (cell.subCells.count > 0) {
+            [mutableArtworks addObjectsFromArray:[self sdl_findAllArtworksToBeUploadedFromCells:cell.subCells]];
+        }
+    }
+
+    return [mutableArtworks copy];
+}
+
+#pragma mark IDs
+
+- (void)sdl_updateIdsOnMenuCells:(NSArray<SDLMenuCell *> *)menuCells parentId:(UInt32)parentId {
+    for (SDLMenuCell *cell in menuCells) {
+        cell.cellId = self.lastMenuId++;
+        cell.parentCellId = parentId;
+        if (cell.subCells.count > 0) {
+            [self sdl_updateIdsOnMenuCells:cell.subCells parentId:cell.cellId];
+        }
+    }
+}
+
+#pragma mark Deletes
+
+- (NSArray<SDLRPCRequest *> *)sdl_deleteCommandsForCells:(NSArray<SDLMenuCell *> *)cells {
+    NSMutableArray<SDLRPCRequest *> *mutableDeletes = [NSMutableArray array];
+    for (SDLMenuCell *cell in cells) {
+        if (cell.subCells == nil) {
+            SDLDeleteCommand *delete = [[SDLDeleteCommand alloc] initWithId:cell.cellId];
+            [mutableDeletes addObject:delete];
+        } else {
+            SDLDeleteSubMenu *delete = [[SDLDeleteSubMenu alloc] initWithId:cell.cellId];
```

Comment:
No `SDLDeleteSubMenu` commands are being sent in the following test case:

#### Test cases
I have two separate menus. 
1. Menu 1 has three root items, one is an `AddCommand`; two are `AddSubMenu`s
1. Menu 2 has one root `AddCommand`. 

#### Test Flow
1. Send menu 1
1. Send menu 2

#### Bug Description
The `AddCommand` in Menu 1 is deleted and replaced with the `AddCommand` from Menu 2. The two `AddSubMenu`s are not deleted.

Reply:
This works for me